lib/plugin.c

```diff
@@ -168,7 +174,235 @@ static struct mptcpd_plugin_ops const *token_to_ops(mptcpd_token_t token)
+
+        size_t const len = strlen(filename);
+
+        /*
+          The last three characters in the filename should be the
+          ".so" extension.
+        */
+        char const *const ext = filename + len - ppos;
+
+        return strcmp(ext, so) == 0;
+}
+
+static void load_plugin(char const *filename)
+{
+        void *const handle = dlopen(filename, RTLD_NOW);
+
+        if (handle == NULL) {
+                l_error("%s", dlerror());
+                return;
+        }
+
+        void *const sym = dlsym(handle, L_STRINGIFY(MPTCPD_PLUGIN_SYM));
+
+        if (sym == NULL) {
+                l_error("%s", dlerror());
+                dlclose(handle);
+                return;
+        }
+
+        struct mptcpd_plugin_desc const *const desc = sym;
+
+        /**
+         * @note We assume that plugin names stored in the plugin
+         *       descriptor will be unique .  Path management events
+         *       could potentially be dispatched to the wrong plugin
+         *       with the same name, otherwise.
+         */
+        // Require a plugin name since we map it plugin operations.
+        if (desc->name == NULL) {
+                l_error("No plugin name specified in %s", filename);
+                dlclose(handle);
+                return;
+        }
+
+        struct plugin_info *const p = l_new(struct plugin_info, 1);
+        p->handle = handle;
+        p->desc   = desc;
+
+        // Register plugin.
+        if (!l_queue_insert(_plugin_infos,
+                            p,
+                            compare_plugin_priority,
+                            NULL)) {
+                /*
+                  We should never get here.  The only way to get here
+                  is if either the l_queue pointer argument,
+                  i.e. _plugin_infos, is NULL or if the comparison
+                  function argument is NULL.
+                */
+                l_error("Unexpected error registering plugin \"%s\"",
+                        filename);
+                l_free(p);
+                dlclose(handle);
+        }
+
+        /*
+          Initialization will be performed after all plugins are
+          loaded to taken into account plugin priority.
+        */
+}
+
+static int load_plugins(char const *dir, struct mptcpd_pm *pm)
+{
+        /**
+         * @note It would be simpler to implement this function in
+         *       terms of @c glob() but that introduces a TOCTOU race
+         *       condition between the @c glob() call and subsequent
+         *       calls to @c dlopen().  The below approach closes to
+         *       the TOCTOU race condition.
+         */
+
+        int const fd = open(dir, O_RDONLY | O_DIRECTORY);
+
+        if (fd == -1) {
+                report_error(errno, "Unable to open plugin directory");
+
+                return -1;
+        }
+
+        DIR *const ds = fdopendir(fd);
+
+        errno = 0;
+        for (struct dirent const *d = readdir(ds);
```

Comment:
Does readdir() tolerate a NULL pointer?

Reply:
According to the [`readdir(3)`](https://www.man7.org/linux/man-pages/man3/readdir.3.html) man page, it should set `errno` to `EBADF` when given a bad directory stream argument.  I can add a `NULL` `DIR*` pointer check if you think that would be better.

Reply:
As long as `readdir(3)` doesn't crash when given a NULL pointer, I think the existing code is fine

Reply:
I confirmed that `readdir()` does indeed crash when passed a `NULL` pointer.  Commit 62b5f9f5551d5b0cd4feebf915f9c1eb859c9db0 adds a NULL pointer check prior to the first call to `readdir()`.